sdk/core/core/test/test_json_builder.c

```diff
@@ -0,0 +1,56 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+#include <az_json_builder.h>
+#include <az_span.h>
+#include <az_span_writer.h>s
+
+#include "./az_test.h"
+
+#include <_az_cfg.h>
+
+AZ_NODISCARD az_result write_json(void * _, az_span_action const write) {
+  (void)_;
+  az_json_builder builder = { 0 };
+
+  AZ_RETURN_IF_FAILED(az_json_builder_init(&builder, write));
+
+  // 0         1         2         3
+  // 01234567890123456789012345678901
+  // {"name":true,"foo":["bar",null]}
+  AZ_RETURN_IF_FAILED(az_json_builder_write(&builder, az_json_value_create_object()));
+
+  AZ_RETURN_IF_FAILED(az_json_builder_write_object_member(
+      &builder, AZ_STR("name"), az_json_value_create_boolean(true)));
+
+  {
+    AZ_RETURN_IF_FAILED(
+        az_json_builder_write_object_member(&builder, AZ_STR("foo"), az_json_value_create_array()));
+    AZ_RETURN_IF_FAILED(
+        az_json_builder_write_array_item(&builder, az_json_value_create_string(AZ_STR("bar"))));
+    AZ_RETURN_IF_FAILED(
+        az_json_builder_write_array_item(&builder, az_json_value_create_null()));
+    AZ_RETURN_IF_FAILED(az_json_builder_write_array_close(&builder));
+  }
+
```

Comment:
Here's an idea but I'm not sure it will completely work:
az_json_builder_write_object_member(&builder, 
   AZ_STR("name"), az_json_value_create_boolean(true)));

// *_array is a varargs function; it automatically adds the [ and the ] at the beginning and end; I just need to list the array items
az_json_builder_write_array(&builder, AZ_STR("foo"), 
   az_json_value_create_string(AZ_STR("bar"))
   az_json_value_create_null())

// We could do the same thing for an object 

As I write this, I see that it only works for scalars and not for items that may be complex objects but maybe there is some idea here that could work for common cases. Maybe if we did something like this:

az_json_builder_write(&builder, AZ_STR("foo"), 
   az_json_array_start(),
   az_json_value_create_string(AZ_STR("bar"))
   az_json_value_create_null(),
   az_json_array_end())

Reply:
I'm ok to add some utility functions that will work as DOM. In the same time, I would like to keep current functions for streaming as well.